contrib/win32/win32compat/misc.c

```diff
@@ -1046,8 +1048,233 @@ readpassphrase(const char *prompt, char *outBuf, size_t outBufLen, int flags)
 	debug3("Expression: %s", expression);
 }
+
+int
+get_machine_domain_name(wchar_t *domain, int size)
+{
+	LPWKSTA_INFO_100 pBuf = NULL;
+	NET_API_STATUS nStatus;
+	LPWSTR pszServerName = NULL;
+
+	nStatus = NetWkstaGetInfo(pszServerName, 100, (LPBYTE *)&pBuf);
+	if (nStatus != NERR_Success) {
+		error("Unable to fetch the machine domain, error:%d\n", nStatus);
+		return 0;
+	}
+
+	debug3("Machine domain:%ls", pBuf->wki100_langroup);
+	wcscpy_s(domain, size, pBuf->wki100_langroup);
+
+	if (pBuf != NULL)
+		NetApiBufferFree(pBuf);
+
+	return 1;
+}
+
+/*
+ * This method will fetch all the groups (listed below) even if the user is indirectly a member.
+ * - Local machine groups
+ * - Domain groups
+ * - global group
+ * - universal groups
+*/
+char **
+getusergroups(const char *user, int *ngroups)
+{	
+	LPGROUP_USERS_INFO_0 local_groups = NULL;
+	LPGROUP_USERS_INFO_0 domain_groups = NULL;
+	LPGROUP_USERS_INFO_0 global_universal_groups = NULL;	
+	DWORD num_local_groups_read = 0;
+	DWORD total_local_groups = 0;
+	DWORD num_domain_groups_read = 0;
+	DWORD total_domain_groups = 0;
+	DWORD num_global_universal_groups_read = 0;
+	DWORD total_global_universal_groups = 0;
+
+	DWORD flags = LG_INCLUDE_INDIRECT;
+	NET_API_STATUS nStatus;
+	wchar_t *w_username = NULL;
+	char *user_domain = NULL;	
+	LPWSTR wszDCName = NULL;
+	char *user_name = malloc(strlen(user)+1);
+	if(!user_name) {
+		error("failed to allocate memory!");
+		return 0;
+	}
+
+	memcpy(user_name, user, strlen(user)+1);
+
+	if (user_domain = strchr(user_name, '@')) {
+		char *t = user_domain;
+		user_domain++;
+		*t='\0';
+	}
+
+	w_username = utf8_to_utf16(user_name);
+	if (!w_username) {
+		error("utf8_to_utf16 failed! for %s", user_name);
+		return NULL;
+	}
+
+	/* Fetch groups on the Local machine */
+	wchar_t w_machine_domain_name[64] = { 0, };
+	if (get_machine_domain_name(w_machine_domain_name, 64)) {
+		if(!machine_domain_name)
+			machine_domain_name = utf16_to_utf8(w_machine_domain_name);
+
+		wchar_t local_user_fmt[256] = { 0, };
+		if (user_domain) {
+			wcscpy_s(local_user_fmt, 256, w_machine_domain_name);
+			wcscat_s(local_user_fmt, 256, L"\\");
+		}
+
+		wcscat_s(local_user_fmt, 256, w_username);
+		nStatus = NetUserGetLocalGroups(NULL,
+			    local_user_fmt,
+			    0,
+			    flags,
+			    (LPBYTE *)&local_groups,
+			    MAX_PREFERRED_LENGTH,
+			    &num_local_groups_read,
+			    &total_local_groups);
+
+		if (NERR_Success != nStatus)
+			error("Failed to get local groups on this machine, error: %d\n", nStatus);
+	}
+	
+	if (user_domain) {
+		/* Fetch Domain groups */
+		nStatus = NetGetDCName(NULL, w_machine_domain_name, (LPBYTE *)&wszDCName);
+		if (NERR_Success == nStatus) {
+			debug3("domain controller name: %ls", wszDCName);
+
+			nStatus = NetUserGetLocalGroups(wszDCName,
+				    w_username,
+				    0,
+				    flags,
+				    (LPBYTE *)&domain_groups,
+				    MAX_PREFERRED_LENGTH,
+				    &num_domain_groups_read,
```

Comment:
same comments. Do we still need the check num_domain_groups_read == total_domain_groups?